src/me/mrCookieSlime/Slimefun/listeners/BlockListener.java

```diff
@@ -75,73 +74,15 @@ public void onPistonRetract(BlockPistonRetractEvent e) {
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.DOWN).getRelative(BlockFace.DOWN), blocks[6]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.DOWN).getRelative(BlockFace.DOWN), blocks[8])
-						&&
-						blocks[0] != null && blocks[0] == blocks[2] 
-						&& BlockAdjacents.hasMaterialOnBothSides(b, blocks[0]) &&
-						blocks[3] != null && blocks[3] == blocks[5] 
-						&& BlockAdjacents.hasMaterialOnBothSides(b.getRelative(0, -1, 0), blocks[5]) &&
-						blocks[6] != null && blocks[6] == blocks[8] 
-						&& BlockAdjacents.hasMaterialOnBothSides(b.getRelative(0, -2, 0), blocks[8])
-						) || (
-						mb.getTriggerBlock() == blocks[4] &&
-						BlockAdjacents.hasMaterialOnSide(b, blocks[3]) &&
-						BlockAdjacents.hasMaterialOnSide(b, blocks[5]) &&
-						BlockAdjacents.isMaterial(b.getRelative(BlockFace.DOWN), blocks[7]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.DOWN), blocks[6]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.DOWN), blocks[8]) &&
-						BlockAdjacents.isMaterial(b.getRelative(BlockFace.UP), blocks[1]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.UP), blocks[0]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.UP), blocks[2])
-						&&
-						blocks[0] != null && blocks[0] == blocks[2]
-						&& BlockAdjacents.hasMaterialOnBothSides(b.getRelative(0, 1, 0), blocks[0]) &&
-						blocks[3] != null && blocks[3] == blocks[5]
-						&& BlockAdjacents.hasMaterialOnBothSides(b, blocks[5]) &&
-						blocks[6] != null && blocks[6] == blocks[8]
-						&& BlockAdjacents.hasMaterialOnBothSides(b.getRelative(0, -1, 0), blocks[8])
-						) || (
-						mb.getTriggerBlock() == blocks[7] &&
-						BlockAdjacents.hasMaterialOnSide(b, blocks[6]) &&
-						BlockAdjacents.hasMaterialOnSide(b, blocks[8]) &&
-						BlockAdjacents.isMaterial(b.getRelative(BlockFace.UP).getRelative(BlockFace.UP), blocks[1]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.UP).getRelative(BlockFace.UP), blocks[0]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.UP).getRelative(BlockFace.UP), blocks[2]) &&
-						BlockAdjacents.isMaterial(b.getRelative(BlockFace.UP), blocks[4]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.UP), blocks[3]) &&
-						BlockAdjacents.hasMaterialOnSide(b.getRelative(BlockFace.UP), blocks[5])
-						&& 
-						blocks[0] != null && blocks[0] == blocks[2]
-						&& BlockAdjacents.hasMaterialOnBothSides(b.getRelative(0, 2, 0), blocks[0]) &&
-						blocks[3] != null && blocks[3] == blocks[5]
-						&& BlockAdjacents.hasMaterialOnBothSides(b.getRelative(0, 1, 0), blocks[5]) &&
-						blocks[6] != null && blocks[6] == blocks[8]
-						&& BlockAdjacents.hasMaterialOnBothSides(b, blocks[8])
-					)) {
-						multiblocks.add(mb);
-					}
+				Player p = e.getPlayer();
+				Block b = e.getClickedBlock();
+				List<MultiBlock> multiblocks = new ArrayList<>();
```

Comment:
I know you didn't design it this way, that was my past self actually.

But you could swap this for a LinkedList, that would make the calls to .add() and retrieving the last element (getLast() or so) a lot cheaper on the CPU.